File: Data Structures & Algorithms/course-schedule-ii/submission-2.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        indegree = [0] * numCourses

        for u, v in prerequisites:
            graph[v].append(u)
            indegree[u] += 1
        
        # queue all nodes with indegree = 0
        q = deque([i for i in range(numCourses) if indegree[i] == 0])

        topo = []

        while q:
            node = q.popleft()
            topo.append(node)

            for nei in graph[node]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)
        
        # need to check if has cycle
        if len(topo) != numCourses:
            return []

        return topo

File: Data Structures & Algorithms/course-schedule/submission-2.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> inorder(numCourses);
        vector<int> topo;

        for (auto& pre : prerequisites) {
            int to = pre[0];
            int from = pre[1];
            graph[from].push_back(to);
            inorder[to]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (inorder[i] == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            topo.push_back(curr);

            for (int nei : graph[curr]) {
                inorder[nei]--;
                if (inorder[nei] == 0) {
                    q.push(nei);
                }
            }
        }

        return topo.size() == numCourses;
    }
};
